crates/nu-protocol/src/pipeline/byte_stream.rs

```diff
@@ -305,33 +445,64 @@ impl ByteStream {
         }
     }
 
-    /// Collect all the bytes of the [`ByteStream`] into a [`String`].
+    /// Collect the stream into a `String` in-memory. This can only succeed if the data contained is
+    /// valid UTF-8.
     ///
-    /// The trailing new line (`\n` or `\r\n`), if any, is removed from the [`String`] prior to being returned.
+    /// The trailing new line (`\n` or `\r\n`), if any, is removed from the [`String`] prior to
+    /// being returned, if this is a stream coming from an external process.
     ///
-    /// If utf-8 decoding fails, an error is returned.
+    /// If the [type](.type_()) is specified as `Binary`, this operation always fails, even if the
+    /// data would have been valid UTF-8.
     pub fn into_string(self) -> Result<String, ShellError> {
         let span = self.span;
-        let bytes = self.into_bytes()?;
-        let mut string = String::from_utf8(bytes).map_err(|_| ShellError::NonUtf8 { span })?;
-        trim_end_newline(&mut string);
-        Ok(string)
+        if self.type_ != ByteStreamType::Binary {
+            let trim = matches!(self.stream, ByteStreamSource::Child(..));
```

Comment:
I would appreciate your opinion here Devyn. I originally had it so that only external command output would be trimmed. But I felt that this should be extended to `File` as well if external command output was trimmed. But then, if file and external streams are trimmed, it would be weird if only `ByteStreamSource::Read` wasn't trimmed. So, I ended up making it trim for all byte streams (e.g., `to text`). Not sure if this is the best option.

This issue doesn't have to be fixed in this PR, but I just noticed that you restored the end line trimming to where I had it in the early stages of the bytestream PR.

Reply:
Oh, yeah, maybe files too then. That makes sense.

Do you think bringing back an explicit boolean flag would be better?

Reply:
Yeah I think a flag would nice to have, I'm just not sure where it would be best to put it. E.g., would it be on the source (`open` and `run-external`)? But then this would require storing and tracking the trim bool (which I guess wouldn't be too bad). Also, the bool might be unused in certain cases like in `^external | str length`. Or, we could add something like a `keep-trailing-newline` flag to `into string`. But then, the flag wouldn't apply to `ByteStreamSource::Read`, since the trailing newline is always kept in this case.

Anyways, just some thoughts, not necessary for this PR. I also think adding trimming to the `File` case would be good, then this comment can be resolved.

Reply:
I think I've handled it now!